src/eigen/preconditioners.jl

```diff
@@ -37,6 +37,8 @@ function PreconditionerTPA(basis::PlaneWaveBasis{T}, kpt::Kpoint; default_shift=
     isempty(kinetic_term) && error("Preconditioner should be disabled when no Kinetic term is used.")
     scaling = only(kinetic_term).scaling_factor
     kin = Vector{T}([scaling * sum(abs2, q) for q in Gplusk_vectors_cart(basis, kpt)] ./ 2)
+    array_type = typeof(similar(basis.G_vectors, T, size(kin)))
+    kin = convert(array_type, kin)
```

Comment:
`map` over Gplusk_vectors_cart (which should be on the GPU)?

Reply:
I'm having some trouble writing `Gplusk_vectors_cart ` as a GPU-compatible function.
Currently, we are broadcasting functions such as `vector_red_to_cart(model::Model, rred)` onto the `G_vectors`. However:
-it is not possible to change these functions to have the signature (recip_lattice::Mat3, rred) as broadcast fails to find common dimensions
-we shouldn't change the function's signature, as sometimes we need the `recip_lattice`, sometimes we need the inverse, and sometimes we need to compute their adjoint
-for GPUs, we can't map or broadcast using a function taking model as an argument as it is not `isbits` (and many of its fields shouldn't be on the GPU anyway).

So far, the best thing I can think of is to replace the functions such as `vector_red_to_cart` into functions returning a static matrix `A` (basically extracting the model's lattice and maybe computing the adjoint) and then mapping the `A*G` operation onto the `G_vectors`, but it's not very satisfying.